std/string.d

```diff
@@ -206,23 +206,73 @@ class StringException : Exception
     Params:
         cString = A null-terminated c-style string.
 
-    Returns: A D-style array of `char` referencing the same string.  The
-    returned array will retain the same type qualifiers as the input.
+    Returns: A D-style array of `char`, `wchar` or `dchar` referencing the same
+    string. The returned array will retain the same type qualifiers as the input.
 
     $(RED Important Note:) The returned array is a slice of the original buffer.
     The original data is not changed and not copied.
 +/
 
-inout(char)[] fromStringz(inout(char)* cString) @nogc @system pure nothrow {
+inout(char)[] fromStringz(inout(char)* cString) @nogc @system pure nothrow
+{
     import core.stdc.string : strlen;
+
     return cString ? cString[0 .. strlen(cString)] : null;
 }
 
+/++ Ditto +/
+inout(Char)[] fromStringz(Char)(inout(Char)* cString) @nogc @system pure nothrow
+if (is(Char == wchar) || is(Char == dchar))
```

Comment:
If we can write`fromStringz!wchar` it would be nice if we could also write `fromStringz!char`. Was there another consideration that led you to keep the `char` version of `fromStringz` as a non-templated overload?

Reply:
@n8sh Backwards compatibility. `fromStringz("foo");` (as per the original unit tests) does not compile with the template version. I tried to come up with a solution that worked as you originally suggested. Perhaps you will have more luck.

Reply:
Honestly, I don't see much reason to care about `fromStringz("foo")`. The entire point of `fromStringz` is to convert from a null-terminated string to a dynamic array. As such, aside from `"foo"` being an easy way to get a null-terminated string for a test, I see no value in accepting it. It would never make sense to pass a string literal to `fromStringz` in real code.

Reply:
@jmdavis True but I would hate to break everyone's code who is using it incorrectly.

Reply:
I would be _very_ surprised if anyone actually uses `fromStringz` with a string literal. It's completely pointless. If you want to turn a string literal into a `string`, just use it as a `string`. It's already a `string`.

Reply:
Will d choose non-deprecated function over a deprecated one during override/template matching?

Reply:
I am happy to change the implementation to a single template function that breaks the API (and unittests) if @jmdavis @wilzbach, and @burner are happy with that. Otherwise I will leave it as is because the current solution doesn't break anything. Or unless @n8sh can suggest a work-around.

Reply:
I'm surprised there was an issue with string literals. Anyway, it doesn't seem necessary to consolidate the existing `fromStringz` into a template if it makes anything more difficult.

Reply:
@n8sh Well if you can get the following to compile (without modifying main)

```d
inout(Char)[] fromStringz(Char = char)(inout(Char)* cString)
if (is(Char == char) || is(Char == wchar) || is (Char == dchar))
{
    import core.stdc.stddef : wchar_t;
    static if (Char.sizeof == 1)
        import core.stdc.string : cstrlen = strlen;
    else static if (Char.sizeof == wchar_t.sizeof)
        import core.stdc.wchar_ : cstrlen = wcslen;
    else
        static size_t cstrlen(const Char* s)
    	{
        	const(Char)* p = s;
       		while (*p)
                ++p;
        	return p - s;
        }

    return cString ? cString[0 .. cstrlen(cString)] : null;
}

void main()
{
    auto a = fromStringz(null);
    auto b = fromStringz("foo");
    auto c = fromStringz("foo\0"c.ptr);
    auto d = fromStringz("foo\0"w.ptr);
    auto d = fromStringz("foo\0"d.ptr);
}
```

Reply:
> Well if you can get the following to compile (without modifying main)

I haven't really followed the discussion, and maybe this falls short in some regard. But string literals convert implicitly to pointers, so you can get that main to compile by providing concrete overloads instead of a template.

For the null literal, provide an overload that takes `typeof(null)`. That seems somewhat pointless to me, though. Why would a user write `fromStringz(null)` instead of `null`, `[]`, or `""`?

```d
inout(Char)[] fromStringzImpl(Char = char)(inout(Char)* cString)
if (is(Char == char) || is(Char == wchar) || is (Char == dchar))
{
    import core.stdc.stddef : wchar_t;
    static if (Char.sizeof == 1)
        import core.stdc.string : cstrlen = strlen;
    else static if (Char.sizeof == wchar_t.sizeof)
        import core.stdc.wchar_ : cstrlen = wcslen;
    else
        static size_t cstrlen(const Char* s)
        {
            const(Char)* p = s;
            while (*p)
                ++p;
            return p - s;
        }

    return cString ? cString[0 .. cstrlen(cString)] : null;
}

alias fromStringz = fromStringzImpl!char;
alias fromStringz = fromStringzImpl!wchar;
alias fromStringz = fromStringzImpl!dchar;

inout(char)[] fromStringz(inout typeof(null) cString)
{
    return fromStringzImpl!char(null);
}

void main()
{
    auto a = fromStringz(null);
    auto b = fromStringz("foo");
    auto c = fromStringz("foo\0"c.ptr);
    auto d = fromStringz("foo\0"w.ptr);
    auto e = fromStringz("foo\0"d.ptr);
}
```

Reply:
@aG0aep6G Thanks for your input, I didn't see your comment until I pushed my solution. I think my version its simpler, but similar. I welcome any comments.